### ISSUE TITLE
HMAC based token pattern scheme

### ISSUE THREAD

**Issue by 69slimm** [OWASP/CheatSheetSeries#76]
Excusme, I wanna ask you about this :

> https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.md

i want to implement this but i still got no idea about the flow work HMAC based token, pointly on verify the token using HMAC. can you explain more detail sir ? Any help from you will be useful for me in the future. Thank you

**Comment by righettod** [comment]
Hi,
We will try to come back to you ASAP in order to help you.

**Comment by 69slimm** [comment]
thanks for the responds, i will waiting.

**Comment by manideepkonakandla** [comment]
Hey @69slimm 

I wrote that HMAC based token method few months back but on re-reading it now, I can see why my explanation was not clear. I will make the explanation more clear on the cheat sheet by end of this week (Basically your token will be a combination of HMAC and couple of other identifiers). I am also going to make it more simplified by removing the operation/data/payload field as well.  

Thank you so much for raising this issue to us!

**Comment by righettod** [comment]
@manideepkonakandla Can I assign this action to you in order that you help @69slimm on this topic ? We can assist you in case of need.
Thanks you very much in advance.

**Comment by 69slimm** [comment]
@manideepkonakandla ok sir. im gonna waiting. Just text me again here sir.

**Comment by manideepkonakandla** [comment]
@righettod  Yes sure, you can assign it to me.

**Comment by righettod** [comment]
GitHub do not let me assign it to @manideepkonakandla because i cannot find him in the assignee list, it's strange, anyone have already meet this case ?
@ThunderSon @mackowski @jmanico

**Comment by jmanico** [comment]
https://github.com/manideepkonakandla looks real to me...

--
Jim Manico
@Manicode
Secure Coding Education
+1 (808) 652-3805

**Comment by righettod** [comment]
Do not work unfortunately...

![NotFound](https://user-images.githubusercontent.com/1573775/56292035-ac5df100-6126-11e9-9df5-758b33a93267.png)

**Comment by manideepkonakandla** [comment]
Not sure why. If you were not able to figure it out, dont worry! I have it on my task list and will complete it over the weekend.

**Comment by righettod** [comment]
Thanks you, we will investigate in parallel to fix the issue...

**Comment by 69slimm** [comment]
@manideepkonakandla @righettod 
hey guys what about this scenario : 
1. i will generate the key of HMAC randomly and random string  then put them as session values
2. And then generate token for hidden field using HMAC algorithm with that key and that random string put it too as session value
And What's next ? How to validate that in web server ?
this is what i think : if the key and random string exist, then calculate them with HMAC algorithm and compare the value with that token in the hidden field. And if true (same) user get the access.

**Comment by manideepkonakandla** [comment]
@69slimm You are missing few things like timestamp which helps you in preventing from replay based attacks. I am sorry for the delay. I just made a PR making the changes. After the review, it will be merged. Once merged, you can clearly see the step-by-step instructions in it (You can see them for now in the commits tab of PR)

https://github.com/OWASP/CheatSheetSeries/pull/78

**Comment by 69slimm** [comment]
Tell me about that "replay based attacks" scenario and how timestamp can protect that ? I am still confusing sir..

**Comment by manideepkonakandla** [comment]
Can you try reading below wiki once? If a token is stolen, it helps in preventing the token usage by the attacker forever (/till the sessionID is changed).

https://en.wikipedia.org/wiki/Replay_attack

**Comment by 69slimm** [comment]
>  it helps in preventing the token 


you mean "it" is timestamp ?

**Comment by manideepkonakandla** [comment]
yes

**Comment by 69slimm** [comment]
thanks for your respond sir. You help me a lot. Maybe in the futur Can i ask you again about all cheat sheet in this github ?

**Comment by manideepkonakandla** [comment]
You can open new issue and one of us will definitely help you :)

**Comment by 69slimm** [comment]
Ok sir thanks Once again

**Comment by righettod** [comment]
Generally for the technical help, we prefer to use the project's Slack channel (see below) or the project's [Google Group](https://groups.google.com/a/owasp.org/forum/#!forum/cheat-sheets-project) but, as @manideepkonakandla said, if you open an issue on the project, we will help you for sure 😃 

Slack channel info:
* Server `owasp.slack.com`
* Channel `cheatsheets`

**Comment by 69slimm** [comment]
Sir Do you know how to do CSRF attack and brute force attack ? My teacher told me to do those action to proof and make comparison table before i do prevention and after prevention ?

**Comment by manideepkonakandla** [comment]
@69slimm I think you should reach out to folks on http://stackoverflow.com for help. We help people if there are any identified issues in the cheat sheets.

**Comment by manideepkonakandla** [comment]
@righettod I think this issue is resolved,  If you feel so as well, can you resolve it ?I dont have the option to do so.

**Comment by ThunderSon** [comment]
It'll be closed once the PR is merged since it references it.

**Comment by righettod** [comment]
Linked to PR #78

**Comment by 69slimm** [comment]
thanks guys.

**Comment by jmanico** [comment]
Train your Developers on Secure Coding

View this email in your browser (https://mailchi.mp/875ff264e773/manicode-security-june-2019?e=1b1f1bde99)


** Manicode Security
------------------------------------------------------------


** Secure Coding Education
June 2019 Update
------------------------------------------------------------
Secure Coding Classes
Thank you for being a customer, partner or associate of Manicode Security.

Manicode Security is now offering over a dozen different application security classes delivered by four different instructors.
* Ron Perris is now teaching classes in React, Node.js, Express and JavaScript secure coding.
* Jimmy Mesta is teaching classes in Kubernetes, AWS, GPC, Cloud Security and DevOps security.
* Dr. Philppe De Ryck is teaching classes in Advanced Web Security.
* And Jim Manico is teaching classes in a wide variety of Web and API Security topics.
* Classes are also available in Mobile Secure Coding and Penetration Testing.

You can download our full catalog here https://manicode.us19.list-manage.com/track/click?u=3b283d304057d06aaba6dc8fb&id=f5e5c7f0ff&e=1b1f1bde99

Please drop us a line at info@manicode.com (mailto:info@manicode.com?subject=Info-June-2019) if you have any questions about our application security classes!
Manicode Security is Partnered with Bit Discovery

The one singular thing literally every information security professional agrees upon is that all organizations must have an inventory of all their internet-connected assets. This includes all websites, name servers, mail servers, IoT devices, blogs, and so on. Of course, an inventory is also required by most every information security standard. And the reason is simple, you cannot secure what you don’t know that you own. Period. Creating a complete asset inventory has proven exceptionally technically challenging, time-consuming, and labor intensive. This has to change. Bit Discovery’s mission, having invested the last several years creating a fantastic product, is making sure every organization has a complete asset inventory. Instantly created. Automatically updated.

Please visit https://manicode.us19.list-manage.com/track/click?u=3b283d304057d06aaba6dc8fb&id=a6b3d01345&e=1b1f1bde99 for more information or email contact@bitdiscovery.com (mailto:contact@bitdiscovery.com)  to schedule a live demo to create an asset inventory for your organization and receive a free trial account.

============================================================

This email was sent to reply@reply.github.com (mailto:reply@reply.github.com)
why did I get this? (https://manicode.us19.list-manage.com/about?u=3b283d304057d06aaba6dc8fb&id=94b725b0f5&e=1b1f1bde99&c=b62483214f)     unsubscribe from this list (https://manicode.us19.list-manage.com/unsubscribe?u=3b283d304057d06aaba6dc8fb&id=94b725b0f5&e=1b1f1bde99&c=b62483214f)     update subscription preferences (https://manicode.us19.list-manage.com/profile?u=3b283d304057d06aaba6dc8fb&id=94b725b0f5&e=1b1f1bde99)
Manicode Security . PO Box 676 . Anahola, HI 96703-0676 . USA

Email Marketing Powered by Mailchimp
http://www.mailchimp.com/monkey-rewards/?utm_source=freemium_newsletter&utm_medium=email&utm_campaign=monkey_rewards&aid=3b283d304057d06aaba6dc8fb&afl=1
